### PR TITLE
Bugfixes, In-app reporting and card support

### PR DESCRIPTION
**Pull request by ellykits**

### CODE REVIEW

File: opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/reporting/register/ReportRegisterPresenterTest.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    @After
+    fun `After every test`() {
+        unmockkAll()
+    }
+
+    @Test
+    fun `Should start monthly report`() {
+        justRun { (view as Activity).startActivity(any()) }
```
Comment (bennsimon) [inline opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/reporting/register/ReportRegisterPresenterTest.kt:30]: is this test complete?

File: opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/reporting/register/ReportRegisterPresenterTest.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        reportRegisterPresenter.startReport(ReportGroup.MONTHLY_REPORTS)
+    }
+
+    @Test
+    fun `Should start annual coverage report`() {
+        justRun { (view as Activity).startActivity(any()) }
```
Comment (bennsimon) [inline opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/reporting/register/ReportRegisterPresenterTest.kt:36]: is this test complete?

File: opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/interactor/ChildJsonFormInteractorTest.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import com.vijay.jsonwizard.constants.JsonFormConstants;
+import com.vijay.jsonwizard.interactors.JsonFormInteractor;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class ChildJsonFormInteractorTest {
+
+    @Test
+    public void testChildFormInteractorMapContainsCustomEditTextWidget() {
```
Comment (bennsimon) [inline opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/interactor/ChildJsonFormInteractorTest.java:12]: i guess the map will always have EDIT_TEXT event without extending ... you can add a check for instance type

File: opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/activity/ChildRegisterActivityTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+
+public class ChildRegisterActivityTest extends BaseActivityUnitTest {
+
+    private ChildRegisterActivity childRegisterActivity;
+    private ActivityController<ChildRegisterActivity> controller;
+
+    @Before
+    public void setUp(){
+        MockitoAnnotations.initMocks(this);
```
Comment (bennsimon) [inline opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/activity/ChildRegisterActivityTest.java:24]: no need for this

File: opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/activity/ChildDetailTabbedActivityTest.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    private ChildDetailTabbedActivity  childDetailTabbedActivity;
+    private ActivityController<ChildDetailTabbedActivity> controller;
+
+
+    @Before
+    public void setUp(){
+        MockitoAnnotations.initMocks(this);
+
+        shadowOf(Looper.getMainLooper()).idle();
```
Comment (bennsimon) [inline opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/activity/ChildDetailTabbedActivityTest.java:32]: do you need this?

File: opensrp-unicef-tunisia/src/main/java/org/smartregister/uniceftunisia/reporting/indicatorposition/IndicatorPositionRepository.kt [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        if (!indicatorPositionPref.isNullOrEmpty() && newVersion > oldVersion
+                && newVersion > indicatorPositionPref.toInt()) {
+            database.run {
+                execSQL("DROP TABLE IF EXISTS indicator_position")
+                createTable(this)
+                delete("sqlite_sequence", "name = ?", arrayOf(Constants.TABLE_NAME))
+            }
+        }
+        val indicatorIndicesJson = Utils.readAssetContents(application.applicationContext,
+                "configs/reporting/indicator-positions.json")
```
Comment (bennsimon) [inline opensrp-unicef-tunisia/src/main/java/org/smartregister/uniceftunisia/reporting/indicatorposition/IndicatorPositionRepository.kt:63]: you can move this to the constants class

File: opensrp-unicef-tunisia/src/main/java/org/smartregister/uniceftunisia/reporting/ReportsDao.kt [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    /**
+     * Returns computed vaccine coverage for the [year]
+     */
+    fun getTargetVaccineCounts(year: Int): List<VaccineCount> {
+        val result = readData(SqlQueries.targetVaccineCountsSql(year)) { cursor: Cursor? ->
+            if (cursor != null && cursor.count > 0) {
+                VaccineCount(
+                        name = getCursorValue(cursor, "name")!!,
```
Comment (bennsimon) [inline opensrp-unicef-tunisia/src/main/java/org/smartregister/uniceftunisia/reporting/ReportsDao.kt:209]: columns names can be put in the constants file

File: opensrp-unicef-tunisia/src/main/java/org/smartregister/uniceftunisia/activity/ChildImmunizationActivity.java [hunk 8]
```diff
@@ -130,18 +124,17 @@ public void updateScheduleDate() {
                 calendar.set(Calendar.HOUR_OF_DAY, 1);
                 long hoursSince1AM = (System.currentTimeMillis() - calendar.getTimeInMillis()) / TimeUnit.HOURS.toMillis(1);
                 if (VaccineSchedulesUpdateJob.isLastTimeRunLongerThan(hoursSince1AM) && !UnicefTunisiaApplication.getInstance().alertUpdatedRepository().findOne(childDetails.entityId())) {
-                    super.updateScheduleDate();
+                    String dobString = Utils.getValue(childDetails.getColumnmaps(), "dob", false);
```
Comment (bennsimon) [inline opensrp-unicef-tunisia/src/main/java/org/smartregister/uniceftunisia/activity/ChildImmunizationActivity.java:127]: move `dob` to a constant

**Review by bennsimon — CHANGES_REQUESTED** [review]

File: opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/reporting/register/ReportRegisterPresenterTest.kt (hunk 1 above)
Comment (ellykits) [reply opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/reporting/register/ReportRegisterPresenterTest.kt:30]: Correct. I have added method to verify that startActivity was indeed called.

File: opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/reporting/register/ReportRegisterPresenterTest.kt (hunk 2 above)
Comment (ellykits) [reply opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/reporting/register/ReportRegisterPresenterTest.kt:36]: Corrected. I have added method to verify that startActivity was indeed called.

File: opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/interactor/ChildJsonFormInteractorTest.java (hunk 3 above)
Comment (ellykits) [reply opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/interactor/ChildJsonFormInteractorTest.java:12]: Noted. Correction made

File: opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/activity/ChildRegisterActivityTest.java (hunk 4 above)
Comment (ellykits) [reply opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/activity/ChildRegisterActivityTest.java:24]: Removed.

File: opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/activity/ChildDetailTabbedActivityTest.java (hunk 5 above)
Comment (ellykits) [reply opensrp-unicef-tunisia/src/test/java/org/smartregister/uniceftunisia/activity/ChildDetailTabbedActivityTest.java:32]: Resolved. This was not required

File: opensrp-unicef-tunisia/src/main/java/org/smartregister/uniceftunisia/reporting/indicatorposition/IndicatorPositionRepository.kt (hunk 6 above)
Comment (ellykits) [reply opensrp-unicef-tunisia/src/main/java/org/smartregister/uniceftunisia/reporting/indicatorposition/IndicatorPositionRepository.kt:63]: :+1:

File: opensrp-unicef-tunisia/src/main/java/org/smartregister/uniceftunisia/activity/ChildImmunizationActivity.java (hunk 8 above)
Comment (ellykits) [reply opensrp-unicef-tunisia/src/main/java/org/smartregister/uniceftunisia/activity/ChildImmunizationActivity.java:127]: Fixed

**Review by bennsimon — APPROVED** [review]

**Pull request merged**